File: fraiseql-go/examples/ecommerce_with_observers.go
package main

import (
	"fmt"
	"log"

	"github.com/fraiseql/fraiseql-go/fraiseql"
)

// Order represents an e-commerce order
type Order struct {
	ID            string  `json:"id"`
	CustomerEmail string  `json:"customer_email"`
	Status        string  `json:"status"`
	Total         float64 `json:"total"`
	CreatedAt     string  `json:"created_at"`
}

// Payment represents a payment record
type Payment struct {
	ID          string   `json:"id"`
	OrderID     string   `json:"order_id"`
	Amount      float64  `json:"amount"`
	Status      string   `json:"status"`
	ProcessedAt *string  `json:"processed_at"`
}

func main() {
	// Register types
	fraiseql.RegisterType("Order", []fraiseql.FieldInfo{
		{Name: "id", Type: "ID", Nullable: false},
		{Name: "customer_email", Type: "String", Nullable: false},
		{Name: "status", Type: "String", Nullable: false},
		{Name: "total", Type: "Float", Nullable: false},
		{Name: "created_at", Type: "DateTime", Nullable: false},
	}, "E-commerce order")

	fraiseql.RegisterType("Payment", []fraiseql.FieldInfo{
		{Name: "id", Type: "ID", Nullable: false},
		{Name: "order_id", Type: "ID", Nullable: false},
		{Name: "amount", Type: "Float", Nullable: false},
		{Name: "status", Type: "String", Nullable: false},
		{Name: "processed_at", Type: "DateTime", Nullable: true},
	}, "Payment record")

	// Observer 1: Notify when high-value orders are created
	fraiseql.NewObserver("onHighValueOrder").
		Entity("Order").
		Event("INSERT").
		Condition("total > 1000").
		Actions(
			fraiseql.Webhook("https://api.example.com/high-value-orders"),
			fraiseql.Slack("#sales", "🎉 High-value order {id}: ${total}"),
			fraiseql.EmailAction(
				"sales@example.com",
				"High-value order {id}",
				"Order {id} for ${total} was created by {customer_email}",
			),
		).
		Register()

	// Observer 2: Notify when orders are shipped
	fraiseql.NewObserver("onOrderShipped").
		Entity("Order").
		Event("UPDATE").
		Condition("status.changed() and status == 'shipped'").
		Actions(
			fraiseql.WebhookWithEnv("SHIPPING_WEBHOOK_URL"),
			fraiseql.EmailAction(
				"{customer_email}",
				"Your order {id} has shipped!",
				"Your order is on its way. Track it here: https://example.com/track/{id}",
				map[string]interface{}{
					"from_email": "noreply@example.com",
				},
			),
		).
		Register()

	// Observer 3: Alert on payment failures with aggressive retry
	fraiseql.NewObserver("onPaymentFailure").
		Entity("Payment").
		Event("UPDATE").
		Condition("status == 'failed'").
		Actions(
			fraiseql.Slack("#payments", "⚠️ Payment failed for order {order_id}: {amount}"),
			fraiseql.Webhook("https://api.example.com/payment-failures", map[string]interface{}{
				"headers": map[string]string{
					"Authorization": "Bearer {PAYMENT_API_TOKEN}",
				},
			}),
		).
		Retry(fraiseql.RetryConfig{
			MaxAttempts:     5,
			BackoffStrategy: "exponential",
			InitialDelayMs:  100,
			MaxDelayMs:      60000,
		}).
		Register()

	// Observer 4: Archive deleted orders
	fraiseql.NewObserver("onOrderDeleted").
		Entity("Order").
		Event("DELETE").
		Action(fraiseql.Webhook("https://api.example.com/archive", map[string]interface{}{
			"body_template": `{"type": "order", "id": "{{id}}", "data": {{_json}}}`,
		})).
		Register()

	// Observer 5: Simple notification for all new orders
	fraiseql.NewObserver("onOrderCreated").
		Entity("Order").
		Event("INSERT").
		Action(fraiseql.Slack("#orders", "New order {id} by {customer_email}")).
		Register()

	// Export schema
	err := fraiseql.ExportSchema("ecommerce_schema.json")
	if err != nil {
		log.Fatalf("Failed to export schema: %v", err)
	}

	fmt.Println("\n🎯 Observer Summary:")
	fmt.Println("   1. onHighValueOrder → Webhooks, Slack, Email for total > 1000")
	fmt.Println("   2. onOrderShipped → Webhook + customer email when status='shipped'")
	fmt.Println("   3. onPaymentFailure → Slack + webhook with retry on payment failures")
	fmt.Println("   4. onOrderDeleted → Archive deleted orders via webhook")
	fmt.Println("   5. onOrderCreated → Slack notification for all new orders")
	fmt.Println("\n✨ Next steps:")
	fmt.Println("   1. fraiseql-cli compile ecommerce_schema.json")
	fmt.Println("   2. fraiseql-server --schema ecommerce_schema.compiled.json")
	fmt.Println("   3. Observers will execute automatically on database changes!")
}

File: fraiseql-php/examples/ecommerce_with_observers.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/../vendor/autoload.php';

use FraiseQL\Attributes\{GraphQLType, GraphQLField};
use FraiseQL\{SchemaRegistry, SchemaFormatter, ObserverBuilder, Webhook, SlackAction, EmailAction};

// Define Order type
#[GraphQLType(name: 'Order', description: 'E-commerce order')]
class Order
{
    #[GraphQLField(type: 'ID')]
    public string $id;

    #[GraphQLField(type: 'String')]
    public string $customerEmail;

    #[GraphQLField(type: 'String')]
    public string $status;

    #[GraphQLField(type: 'Float')]
    public float $total;

    #[GraphQLField(type: 'DateTime')]
    public string $createdAt;
}

// Define Payment type
#[GraphQLType(name: 'Payment', description: 'Payment record')]
class Payment
{
    #[GraphQLField(type: 'ID')]
    public string $id;

    #[GraphQLField(type: 'ID')]
    public string $orderId;

    #[GraphQLField(type: 'Float')]
    public float $amount;

    #[GraphQLField(type: 'String')]
    public string $status;

    #[GraphQLField(type: 'DateTime', nullable: true)]
    public ?string $processedAt;
}

// Register types
$registry = SchemaRegistry::getInstance();
$registry->register(Order::class);
$registry->register(Payment::class);

// Observer 1: High-value orders
ObserverBuilder::create('onHighValueOrder')
    ->entity('Order')
    ->event('INSERT')
    ->condition('total > 1000')
    ->addAction(Webhook::create('https://api.example.com/high-value-orders'))
    ->addAction(SlackAction::create('#sales', '🎉 High-value order {id}: ${total}'))
    ->addAction(EmailAction::create(
        'sales@example.com',
        'High-value order {id}',
        'Order {id} for ${total} was created by {customer_email}'
    ))
    ->register();

// Observer 2: Order shipped
ObserverBuilder::create('onOrderShipped')
    ->entity('Order')
    ->event('UPDATE')
    ->condition("status.changed() and status == 'shipped'")
    ->addAction(Webhook::withEnv('SHIPPING_WEBHOOK_URL'))
    ->addAction(EmailAction::withFrom(
        '{customer_email}',
        'Your order {id} has shipped!',
        'Your order is on its way. Track it here: https://example.com/track/{id}',
        'noreply@example.com'
    ))
    ->register();

// Observer 3: Payment failures
ObserverBuilder::create('onPaymentFailure')
    ->entity('Payment')
    ->event('UPDATE')
    ->condition("status == 'failed'")
    ->addAction(SlackAction::create('#payments', '⚠️ Payment failed for order {order_id}: {amount}'))
    ->addAction(Webhook::create('https://api.example.com/payment-failures', [
        'headers' => ['Authorization' => 'Bearer {PAYMENT_API_TOKEN}']
    ]))
    ->retry(['max_attempts' => 5, 'backoff_strategy' => 'exponential'])
    ->register();

// Observer 4: Archive deleted orders
ObserverBuilder::create('onOrderDeleted')
    ->entity('Order')
    ->event('DELETE')
    ->addAction(Webhook::create('https://api.example.com/archive', [
        'body_template' => '{"type": "order", "id": "{{id}}", "data": {{_json}}}'
    ]))
    ->register();

// Observer 5: All new orders
ObserverBuilder::create('onOrderCreated')
    ->entity('Order')
    ->event('INSERT')
    ->addAction(SlackAction::create('#orders', 'New order {id} by {customer_email}'))
    ->register();

// Export schema
$formatter = new SchemaFormatter();
$schema = $formatter->formatRegistry($registry);
$schema->saveToFile('ecommerce_observers_schema.json');

echo "\n✅ Schema exported to ecommerce_observers_schema.json\n";
echo "   Types: " . $schema->getTypeCount() . "\n";
echo "   Observers: " . count($schema->observers) . "\n";

echo "\n🎯 Observer Summary:\n";
echo "   1. onHighValueOrder → Webhooks, Slack, Email for total > 1000\n";
echo "   2. onOrderShipped → Webhook + customer email when status='shipped'\n";
echo "   3. onPaymentFailure → Slack + webhook with retry on payment failures\n";
echo "   4. onOrderDeleted → Archive deleted orders via webhook\n";
echo "   5. onOrderCreated → Slack notification for all new orders\n";

echo "\n✨ Next steps:\n";
echo "   1. fraiseql-cli compile ecommerce_observers_schema.json\n";
echo "   2. fraiseql-server --schema ecommerce_observers_schema.compiled.json\n";
echo "   3. Observers will execute automatically on database changes!\n";
